### PR TITLE
Feature: support defining kspacing for a/b/c separately

### PR DESCRIPTION
**Pull request by pxlxingliang**

### CODE REVIEW

File: source/module_io/input.cpp [hunk 1]
```diff
@@ -688,7 +688,35 @@ bool Input::Read(const std::string &fn)
         }
         else if (strcmp("kspacing", word) == 0)
         {
-            read_value(ifs, kspacing);
+            std::string s;
+            std::getline(ifs,s);
+            std::stringstream ss(s);
+
+            //read 3 values
```
Comment (dyzheng) [inline source/module_io/input.cpp]: line 691 - 719 should refactor to a functional and add UnitTest for this function independent.

File: source/module_io/input.cpp [hunk 2]
```diff
@@ -688,7 +688,35 @@ bool Input::Read(const std::string &fn)
...
+                kspacing[1] = kspacing[0];
+                kspacing[2] = kspacing[0];
+            }
+            //std::cout << "count: " << count << " kspacing: " << kspacing[0] << " " << kspacing[1] << " " << kspacing[2] << std::endl; 
+
+            //if has kspacing equal to 0, quit.
+            for (int i=0;i<3;i++)
+            {
+                if (kspacing[i] <= 0){
+                    ModuleBase::WARNING_QUIT("Input", "kspacing should be > 0");
```
Comment (dyzheng) [inline source/module_io/input.cpp]: Don't WARNING_QUIT during reading.

File: source/module_io/input.cpp [hunk 3]
```diff
@@ -3163,9 +3192,11 @@ void Input::Check(void)
     {
         ModuleBase::WARNING_QUIT("Input", "please don't set diago_proc with lcao base");
     }
-    if (kspacing < 0.0)
-    {
-        ModuleBase::WARNING_QUIT("Input", "kspacing must > 0");
+    for (int i=0;i<3;i++){
```
Comment (dyzheng) [inline source/module_io/input.cpp:3168]: maybe you can assert :
```
double maximum=0, multiply=1;
for(i=0;i<3;i++)
{
    maximum = std::max(std::abs(kspacing[i], maximum));
    multiply *= kspacing[i];
    if (kspacing[i] < 0.0)
    {
        ModuleBase::WARNING_QUIT("Input", "kspacing must > 0");
    }
}
if(maximum > small && multiple < small)
{WARNING_QUIT();}
```

**Comment by sunliang98** [comment]
Maybe it's better to add an integrate test with defining kspacing for a/b/c separately.

**Comment by dyzheng** [comment]
> Maybe it's better to add an integrate test with defining kspacing for a/b/c separately.

It is okay to add Integrate test, but UnitTest is better than Integrate test.

File: source/module_io/input.cpp (hunk 2 above)
Comment (hongriTianqi) [reply source/module_io/input.cpp]: Agree: checking of input parameters should be put in the function Input::Check()

File: source/module_cell/klist.cpp [hunk 4]
```diff
@@ -175,16 +175,16 @@ bool K_Vectors::read_kpoints(const std::string &fn)
 		ofs << "1 1 1 0 0 0" << std::endl;
 		ofs.close();
 	}
-    else if (GlobalV::KSPACING > 0.0)
+    else if (GlobalV::KSPACING[0] > 0.0)
```
Comment (hongriTianqi) [inline source/module_cell/klist.cpp:178]: Should we make sure that GlobalV::KSPACING[1] > 0.0 and Global::KSPACING[2] > 0.0?

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_cell/klist.cpp [hunk 5]
```diff
@@ -175,16 +175,16 @@ bool K_Vectors::read_kpoints(const std::string &fn)
...
         ModuleBase::Matrix3 btmp = GlobalC::ucell.G;
         double b1 = sqrt(btmp.e11 * btmp.e11 + btmp.e12 * btmp.e12 + btmp.e13 * btmp.e13);
         double b2 = sqrt(btmp.e21 * btmp.e21 + btmp.e22 * btmp.e22 + btmp.e23 * btmp.e23);
         double b3 = sqrt(btmp.e31 * btmp.e31 + btmp.e32 * btmp.e32 + btmp.e33 * btmp.e33);
-        int nk1 = max(1,static_cast<int>(b1 * ModuleBase::TWO_PI / GlobalV::KSPACING / GlobalC::ucell.lat0 + 1));
-        int nk2 = max(1,static_cast<int>(b2 * ModuleBase::TWO_PI / GlobalV::KSPACING / GlobalC::ucell.lat0 + 1));
-        int nk3 = max(1,static_cast<int>(b3 * ModuleBase::TWO_PI / GlobalV::KSPACING / GlobalC::ucell.lat0 + 1));
+        int nk1 = max(1,static_cast<int>(b1 * ModuleBase::TWO_PI / GlobalV::KSPACING[0] / GlobalC::ucell.lat0 + 1));
+        int nk2 = max(1,static_cast<int>(b2 * ModuleBase::TWO_PI / GlobalV::KSPACING[1] / GlobalC::ucell.lat0 + 1));
+        int nk3 = max(1,static_cast<int>(b3 * ModuleBase::TWO_PI / GlobalV::KSPACING[2] / GlobalC::ucell.lat0 + 1));
```
Comment (hongriTianqi) [inline source/module_cell/klist.cpp:187]: The corresponding unit test for K_Vectors::read_kpoints should be added for this functionality.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

**Review by dyzheng — APPROVED** [review]

**Review by hongriTianqi — APPROVED** [review]

**Pull request merged**